src/scale_impls/encode.rs

```diff
@@ -209,21 +240,49 @@ fn encode_composite<T>(
     Err(original_error)
 }
 
-// skip into the target type past any newtype wrapper like things:
-fn find_single_entry_with_same_repr(type_id: u32, types: &PortableRegistry) -> u32 {
-    let Some(ty) = types.resolve(type_id) else { return type_id };
-    match &ty.type_def {
-        TypeDef::Tuple(tuple) if tuple.fields.len() == 1 => {
-            find_single_entry_with_same_repr(tuple.fields[0].id, types)
-        }
-        TypeDef::Composite(composite) if composite.fields.len() == 1 => {
-            find_single_entry_with_same_repr(composite.fields[0].ty.id, types)
-        }
-        TypeDef::Array(arr) if arr.len == 1 => {
-            find_single_entry_with_same_repr(arr.type_param.id, types)
-        }
-        _ => type_id,
-    }
+/// Skip into the target type past any newtype wrapper like things.
+/// Also returns a bool indicating whether we skipped into something or not.
+/// This bool is true when the returned id is different from the id that was passed in.
+fn find_single_entry_with_same_repr<'a, R: TypeResolver>(
+    type_id: &'a R::TypeId,
+    types: &'a R,
+) -> Result<(&'a R::TypeId, bool), R::Error> {
+    let return_unchanged = (type_id, false);
+    let visitor = scale_type_resolver::visitor::new((), |_, _| return_unchanged)
+        .visit_tuple(|_, fields| {
+            if fields.len() == 1 {
+                let ty = fields.next().expect("has 1 item; qed;");
+                match find_single_entry_with_same_repr(ty, types) {
+                    Ok((inner, _)) => (inner, true),
+                    Err(_) => return_unchanged,
+                }
```

Comment:
I'm not sure if it's right to replace the error on the `Err` cases with the original result; I'd lean towards we should return any error we encounter.

You could also about the matches and needing to remember to replace the returned inner bool with this:

```rust
fn find_single_entry_with_same_repr<'a, R: TypeResolver>(
    type_id: &'a R::TypeId,
    types: &'a R,
) -> Result<(&'a R::TypeId, bool), R::Error> {
    fn do_find<'a, R: TypeResolver>(
        type_id: &'a R::TypeId,
        types: &'a R,
        type_id_has_changed: bool
    ) -> Result<(&'a R::TypeId, bool), R::Error> {
        let return_unchanged = (type_id, type_id_has_changed);
        let visitor = scale_type_resolver::visitor::new((), |_, _| return_unchanged)
            .visit_tuple(|_, fields| {
                if fields.len() == 1 {
                    let ty = fields.next().expect("has 1 item; qed;");
                    do_find(ty, types, true)
                } else {
                    return_unchanged
                }
            })
            .visit_composite(|_, fields| {
                if fields.len() == 1 {
                    let ty = fields.next().expect("has 1 item; qed;").id;
                    do_find(ty, types, true)
                } else {
                    return_unchanged
                }
            })
            .visit_array(|_, ty_id, len| {
                if len == 1 {
                    do_find(ty_id, types, true)
                } else {
                    return_unchanged
                }
            });
    
        types.resolve_type(type_id, visitor)
    }

    do_find(type_id, types, false)
}
```

```rust

```

Reply:
Ooh I see; I was thrown because the return type turned into `Result<..>`, but it appeared like an error was never being returned.

I guess it doesn't matter if we ignore the error, since it'll just lead to the outer type ID being returned and the error will be encountered later on most probably anyways. So somehting like this I guess also works:

```rust
fn find_single_entry_with_same_repr<'a, R: TypeResolver>(
    type_id: &'a R::TypeId,
    types: &'a R,
) -> (&'a R::TypeId, bool) {
    fn do_find<'a, R: TypeResolver>(
        type_id: &'a R::TypeId,
        types: &'a R,
        type_id_has_changed: bool
    ) -> (&'a R::TypeId, bool) {
        let return_unchanged = (type_id, type_id_has_changed);
        let visitor = scale_type_resolver::visitor::new((), |_, _| return_unchanged)
            .visit_tuple(|_, fields| {
                if fields.len() == 1 {
                    let ty = fields.next().expect("has 1 item; qed;");
                    do_find(ty, types, true)
                } else {
                    return_unchanged
                }
            })
            .visit_composite(|_, fields| {
                if fields.len() == 1 {
                    let ty = fields.next().expect("has 1 item; qed;").id;
                    do_find(ty, types, true)
                } else {
                    return_unchanged
                }
            })
            .visit_array(|_, ty_id, len| {
                if len == 1 {
                    do_find(ty_id, types, true)
                } else {
                    return_unchanged
                }
            });
    
        types.resolve_type(type_id, visitor).unwrap_or(return_unchanged)
    }

    do_find(type_id, types, false)
}
```

(untested!)